### ISSUE TITLE
dbgmsg does nothing useful with tail -f

### ISSUE THREAD

**Issue by rincebrain** [openzfs/zfs#12363]
### Describe the feature would like to see added to OpenZFS
Currently, `tail -f /proc/spl/kstat/zfs/dbgmsg` on Linux, rather than just outputting the last N lines then more as they come in, appears to spit out every single line presently in the buffer, then waits forever, never* printing more lines. (I have, admittedly, not tested "never" longer than a few minutes after other readers to dbgmsg can see messages, and am just speculating.)

It would be nice if it would either convince tail to exit with an error (so you know it doesn't do that), or (ideally) feed new lines to a reader as they arrive - and in either case, probably not cause tail -f to spit out every line currently in the buffer.

### How will this feature improve OpenZFS?
A nicer experience for anyone using the /proc/ dbgmsg interface.

### Additional context
I was debugging, and annoyed that I had to remember to work around dbgmsg not working as desired, but haven't gotten around to trying to implement it, so I'm leaving this here in case anyone else cares enough to take a stab before I do.

**Comment by behlendorf** [comment]
This behavior is result of virtual `/proc` files never being designed to work with `tail -f`.  For the curious there's an [lkml thread](https://lkml.org/lkml/2001/10/10/152) from way back in 2001 which discusses why this is the case.  Those fundamental reasons remain true to this day so for this to work we'd need to use an interface other than `/proc/` to get the debug log.  Which actually might be nice since if we did it with an `ioctl()` it'd be more portable and we could add things like `-f` (follow).  It was originally wired up to `/proc` simply because it was convenient and has worked well enough it was never revisited.

**Comment by rincebrain** [comment]
Huh. I swore I recalled once implementing it, but digging out the ancient code I was thinking of, I was misrecalling.

How annoying, thank you for the pointer.

**Comment by rincebrain** [comment]
For anyone stumbling over this thread, I personally solved my immediate-term need with this unpolished systemtap script:
```
%{
#include <linux/time.h>
%}

function eviltime:long() %{
        struct timespec64 ts;
        ktime_get_coarse_real_ts64(&ts);
        STAP_RETURN(ts.tv_sec);
%}

probe kprobe.module("zfs").function("__zfs_dbgmsg").call {
        printf("%ld   %s\n", eviltime(), kernel_string(pointer_arg(1)))
}
```

It hasn't burned the system down yet.

**Comment by youzhongyang** [comment]
bpftrace can do it too, the only limitation is the 200 max chars.
```
env BPFTRACE_STRLEN=200 bpftrace -e 'kprobe:__zfs_dbgmsg { time("%H:%M:%S"); printf(" %s\n", str(arg0)); }'
```

**Comment by rincebrain** [comment]
Yeah, I know, I spent a bit of time trying to find a non-hacky workaround to the 200 char limit in bpftrace first before giving up and just going for stap.
